spmd/api.py

```diff
@@ -0,0 +1,329 @@
+
+import os
+from functools import partial
+from typing import Any, Callable, Dict, List, Sequence, Tuple
+
+
+@dataclass
+class Schema:
+    mesh: DeviceMesh
+    placements: List[Placement]
+
+
+def _dispatch_with_local_tensors(
+    op: torch._ops.OpOverload,
+    local_args: Tuple[object, ...],
+    kwargs: Dict[str, object] = {},
+    specs: Dict[
+        torch.Tensor,
+        Tuple[torch.Size, DeviceMesh, Sequence[Placement], Sequence[Placement]],
+    ] = {},
+) -> Any:
+    def redistribute(arg):
+        return (
+            _redistribute_with_local_tensor(arg, *specs[arg])
+            if arg in specs
+            else arg
+        )
+
+    return op(*tree_map(redistribute, local_args), **kwargs)
+
+
+def _get_dtensor_dispatch_graph(
+    node: fx.Node, node_to_obj: Dict[fx.Node, object],
+) -> fx.GraphModule:
+
+    def remap_arg(arg):
+        if isinstance(arg, torch.fx.Node):
+            obj = node_to_obj[arg]
+            if _get_tracer(obj):
+                # This is a shared arg, already has a tracer from previous
+                # tracing. Delete the tracer.
+                del obj.__dict__[proxy_slot]
+            return obj
+        else:
+            return arg
+
+    print("=== tracing ", node.name, node.target, node.args, node.kwargs)
+    args = tree_map(remap_arg, node.args)
+    # kwargs in this set of tests are all constants
+    kwargs = node.kwargs
+
+    # run dispatch once to get the real DTensor output
+    out = operator_dispatch(
+        node.target,
+        args,
+        node.kwargs,  # kwargs in this set of tests are all constants
+        DTensor._op_to_rules,
+        DTensor._custom_dispatch_ops,
+    )
+    node_to_obj[node] = out
+    print("=== dtensor out, ", out.placements, out.size())
+
+    # get DTensor specs for inputs and outputs
+    (
+        target_schema,
+        redistribute,
+        output_sharding,
+    ) = propagate_input_sharding(
+        node.target,
+        args,
+        kwargs,
+        DTensor._op_to_rules,
+    )
+
+    flatten_args, args_tree_spec = tree_flatten(args)
+    flatten_args_schema, _ = tree_flatten(target_schema.args_schema)
+
+    specs: Dict[
+        torch.Tensor,
+        Tuple[
+            torch.Size,
+            DeviceMesh,
+            Sequence[Placement],
+            Sequence[Placement],
+        ],
+    ] = {}
+    for i, arg in enumerate(flatten_args):
+        if isinstance(arg, DTensor) and redistribute:
+            specs[arg._local_tensor] = (
+                arg.size(),
+                flatten_args_schema[i].mesh,
+                arg.placements,
+                flatten_args_schema[i].placements,
+            )
+
+    dispatch = partial(
+        _dispatch_with_local_tensors,
+        node.target,
+        kwargs=kwargs,
+        specs=specs,
+    )
+
+    def unwrap_local(e):
+        return e._local_tensor if isinstance(e, DTensor) else e
+
+    return make_fx(dispatch)(tree_map(unwrap_local, args))
+
+
+
+class SPMD(nn.Module):
```

Comment:
@aazzolini also asked for this. I extract the main logic to a `_convert_to_distributed` function API. 

Do we still plan to have a module wrapper in tau? I thought that's the original goal?

Reply:
I don't have a strong opinion here, we can iterate on the APIs and finalized it later. My point was that if we are thinking a functional like API for other module wrappers like DDP, it might worth considering we just directly implement a functional like API in tau?